mailbox/api/src/test/java/org/apache/james/mailbox/events/EventDeadLettersContract.java

```diff
@@ -0,0 +1,405 @@
+    Group NULL_GROUP = null;
+    Event NULL_EVENT = null;
+    Event.EventId NULL_EVENT_ID = null;
+
+    EventDeadLetters eventDeadLetters();
+
+    default Stream<Event.EventId> allEventIds() {
+        EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+        return eventDeadLetters.groupsWithFailedEvents()
+            .flatMap(eventDeadLetters::failedEventIds)
+            .toStream();
+    }
+
+    interface StoreContract extends EventDeadLettersContract {
+
+        @Test
+        default void storeShouldThrowWhenNullGroup() {
+            EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+            assertThatThrownBy(() -> eventDeadLetters.store(NULL_GROUP, EVENT_1))
+                .isInstanceOf(IllegalArgumentException.class);
+        }
+
+        @Test
+        default void storeShouldThrowWhenNullEvent() {
+            EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+            assertThatThrownBy(() -> eventDeadLetters.store(GROUP_A, NULL_EVENT))
+                .isInstanceOf(IllegalArgumentException.class);
+        }
+
+        @Test
+        default void storeShouldThrowWhenBothGroupAndEventAreNull() {
+            EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+            assertThatThrownBy(() -> eventDeadLetters.store(NULL_GROUP, NULL_EVENT))
+                .isInstanceOf(IllegalArgumentException.class);
+        }
+
+        @Test
+        default void storeShouldStoreGroupWithCorrespondingEvent() {
+            EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+            eventDeadLetters.store(GROUP_A, EVENT_1).block();
+            assertThat(eventDeadLetters.failedEventId(GROUP_A, EVENT_1.getEventId()).block())
+                .isEqualTo(EVENT_ID_1);
+        }
+
+        @Test
+        default void storeShouldIgnoreStoreDuplicatedEventsPerGroup() {
+            EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+            eventDeadLetters.store(GROUP_A, EVENT_1).block();
+            eventDeadLetters.store(GROUP_A, EVENT_1).block();
+            eventDeadLetters.store(GROUP_A, EVENT_1).block();
+
+            assertThat(eventDeadLetters.failedEventIds(GROUP_A).toStream())
+                .containsExactly(EVENT_ID_1);
+        }
+
+        @Test
+        default void storeShouldKeepConsistencyWhenConcurrentStore() throws Exception {
+            EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+            ConcurrentHashMap<Integer, Group> groups = concurrentGroups();
+            Multimap<Integer, Event.EventId> storedEventIds = Multimaps.synchronizedSetMultimap(HashMultimap.create());
+            runConcurrent((threadNumber, step) -> {
+                Event.EventId eventId = Event.EventId.random();
+                storedEventIds.put(threadNumber, eventId);
+                eventDeadLetters.store(groups.get(threadNumber), event(eventId));
+            });
+
+            groups.forEach((groupId, group) -> {
```

Comment:
`groups.values().forEach(group -> ...)`

Reply:
in bellow lines, I need to use `groupId`  also